examtool/examtool/api/utils.py

```diff
@@ -28,3 +25,44 @@ def dict_to_list(d):
 
 def list_to_dict(l):
     return {i: x for i, x in enumerate(l)}
+
+
+class IDFactory:
+    def __init__(self, id_start="", id_end="", length=32, allow_random_ids=True):
+        self.length = length
+        self.id_start = id_start
+        self.id_end = id_end
+        self.allow_random_ids = allow_random_ids
+        self.current_ids = set()
+        self.lock = Lock()
+
+    def rand_id(self, length=None):
+        if length is None:
+            length = self.length
+        return "".join(random.choices(string.ascii_uppercase, k=length))
+
+    def id_from_str(self, string):
+        return self.id_start + string + self.id_end
+
+    def get_id(self, string=None):
+        if string is None:
+            if not self.allow_random_ids:
+                raise SyntaxError(
+                    "A custom ID is required but was not set for this question!"
+                )
+            qid = self.rand_id()
+            self.lock.acquire()
```

Comment:
Is this class used in the multithreaded part? I thought you're only multithreading the conversion function, which just calls the pandoc library.

If not, no need for locks (we don't need thread safety everywhere, we do stuff like changing the random seed elsewhere anyway which is _definitely_ not thread safe lmao)

Reply:
You are right that it is not handled in there. I am just being stupid. I had not realized where it was happening at first and forgot to revert that change.